sdk/core/core-lro/src/lroEngine/operation.ts

```diff
@@ -0,0 +1,103 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+import { AbortSignalLike } from "@azure/abort-controller";
+import { PollOperationState, PollOperation } from "../pollOperation";
+import {
+  PollerConfig,
+  ResumablePollOperationState,
+  LongRunningOperation,
+  LroStatus
+} from "./models";
+import { getPollingUrl } from "./requestUtils";
+import { createInitializeState, createPollForLROStatus } from "./stateMachine";
+
+export class GenericPollOperation<TResult, TState extends PollOperationState<TResult>>
+  implements PollOperation<TState, TResult> {
+  private getLROStatusFromResponse?: (
+    pollingURL: string,
+    pollerConfig: PollerConfig
+  ) => Promise<LroStatus<TResult>>;
+  private pollerConfig?: PollerConfig;
+  constructor(
+    public state: TState & ResumablePollOperationState<TResult>,
+    private lro: LongRunningOperation<TResult>
+  ) {}
+
+  public setPollerConfig(pollerConfig: PollerConfig): void {
+    this.pollerConfig = pollerConfig;
+  }
+
+  /**
+   * General update function for LROPoller, the general process is as follows
+   * 1. Check initial operation result to determine the strategy to use
+   *  - Strategies: Location, Azure-AsyncOperation, Original Uri
+   * 2. Check if the operation result has a terminal state
+   *  - Terminal state will be determined by each strategy
+   *  2.1 If it is terminal state Check if a final GET request is required, if so
+   *      send final GET request and return result from operation. If no final GET
+   *      is required, just return the result from operation.
+   *      - Determining what to call for final request is responsibility of each strategy
+   *  2.2 If it is not terminal state, call the polling operation and go to step 1
+   *      - Determining what to call for polling is responsibility of each strategy
+   *      - Strategies will always use the latest URI for polling if provided otherwise
+   *        the last known one
+   */
+  async update(options?: {
+    abortSignal?: AbortSignalLike | undefined;
+    fireProgress?: ((state: TState) => void) | undefined;
+  }): Promise<PollOperation<TState, TResult>> {
+    const state = this.state;
+    if (!state.isStarted) {
+      const initializeState = createInitializeState(
+        state,
+        this.lro.requestPath,
+        this.lro.requestMethod
+      );
+      await this.lro.sendInitialRequest(initializeState);
+    }
+
+    if (!state.isCompleted) {
+      if (this.getLROStatusFromResponse === undefined) {
+        if (state.config === undefined) {
+          throw new Error(
+            "Bad state: LRO mode is undefined. Please check if the serialized state is well-formed."
```

Comment:
do you mind if we move the error messages outside of this function into a common place? could be right above the function. I think it will make it easier to keep these messages on sync over time.

Reply:
I did some edits there but felt like pulling out stuff will not help make this more readable.